### PR TITLE
[15.0][MIG] web_timeline: Migration to 15.0 

### PR DESCRIPTION
**Pull request by cesar-tecnativa**
cc @Tecnativa TT35806

### CODE REVIEW

**Comment by cesar-tecnativa** [comment]
@CarlosRoca13 please review

**Comment by cesar-tecnativa** [comment]
@Tardo please review

**Comment by pedrobaeza** [comment]
Have you tested it in a real case like with `project_timeline`?

**Comment by cesar-tecnativa** [comment]
> Have you tested it in a real case like with `project_timeline`?

This is the goal https://github.com/OCA/project/pull/942

File: web_timeline/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    "application": False,
+    "installable": True,
+    "web.assets_backend": [
+        "web_timeline/static/src/scss/web_timeline.scss",
+        "web_timeline/static/src/js/timeline_view.js",
+        "web_timeline/static/src/js/timeline_renderer.js",
+        "web_timeline/static/src/js/timeline_controller.js",
+        "web_timeline/static/src/js/timeline_model.js",
+        "web_timeline/static/src/js/timeline_canvas.js",
+    ],
```
Comment (CarlosRoca13) [inline web_timeline/__manifest__.py]:
```suggestion
    "assets": {
        "web.assets_backend": [
            "web_timeline/static/src/scss/web_timeline.scss",
            "web_timeline/static/src/js/timeline_view.js",
            "web_timeline/static/src/js/timeline_renderer.js",
            "web_timeline/static/src/js/timeline_controller.js",
            "web_timeline/static/src/js/timeline_model.js",
            "web_timeline/static/src/js/timeline_canvas.js",
        ]
    },
```

**Review by CarlosRoca13 — CHANGES_REQUESTED** [review]

**Review by CarlosRoca13 — APPROVED** [review]

**Comment by cesar-tecnativa** [comment]
@sergio-teruel plesae review

**Comment by cesar-tecnativa** [comment]
@LudLaf please review

File: web_timeline/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    "installable": True,
+    "assets": {
+        "web.assets_backend": [
+            "web_timeline/static/src/scss/web_timeline.scss",
+            "web_timeline/static/src/js/timeline_view.js",
+            "web_timeline/static/src/js/timeline_renderer.js",
+            "web_timeline/static/src/js/timeline_controller.js",
+            "web_timeline/static/src/js/timeline_model.js",
+            "web_timeline/static/src/js/timeline_canvas.js",
+        ]
```
Comment (jguenat) [inline web_timeline/__manifest__.py]:
```suggestion
        ],
        "web.assets_qweb": [
            "web_timeline/static/src/xml/web_timeline.xml",
        ],
```

File: web_timeline/static/src/js/timeline_view.js [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    var TimelineView = AbstractView.extend({
+        display_name: _lt("Timeline"),
+        icon: "fa-tasks",
+        jsLibs: ["/web_timeline/static/lib/vis-timeline/vis-timeline-graph2d.js"],
+        cssLibs: ["/web_timeline/static/lib/vis-timeline/vis-timeline-graph2d.css"],
+        config: _.extend({}, AbstractView.prototype.config, {
+            Model: TimelineModel,
+            Controller: TimelineController,
+            Renderer: TimelineRenderer,
+        }),
```
Comment (jguenat) [inline web_timeline/static/src/js/timeline_view.js:33]:
```suggestion
        }),
        viewType: 'timeline',
```

**Review by jguenat — COMMENTED** [review]
Hello,
I tested this with project_timeline.
As discussed there, Group by many2many is not supported by the timeline view.
Anyone know if it's easy to fix?

**Comment by jguenat** [comment]
Also, what do you think about updating vis-timeline-graph2d lib to V 7.7?
https://unpkg.com/vis-timeline@7.7.0/peer/umd/vis-timeline-graph2d.min.js
https://unpkg.com/vis-timeline@7.7.0/styles/vis-timeline-graph2d.min.css

**Comment by chrisandrewmann** [comment]
Issue with grouping by many2many fields causes crash as mentioned here.
https://github.com/OCA/project/pull/942#issuecomment-1198966193

**Comment by chrisandrewmann** [comment]
@CarlosRoca13 Thanks for that, awesome job with adding the m2m field grouping. I had tried myself but javascript is not my strongest point.

**Comment by CarlosRoca13** [comment]
Hi! I added the changes added on https://github.com/OCA/web/pull/2256

I have also corrected the implementation to allow grouping by m2m fields with some restrictions:
When grouping by m2m and more than one record is set, the timeline item appears just on one group.
When grouping by m2m and dragging for changing the time or the group, the changes on the group will not be set, because it could make disappear the records not related with the changes that we want to make.

**Comment by CarlosRoca13** [comment]
I also tried to add the new libraries, but more changes to adjust the code were needed,  so I have chosen to focus on the main problem

**Comment by chrisandrewmann** [comment]
@CarlosRoca13 Have given it a quick test and only problem I see is that any M2M field (i.e Assignees on project.task) that have no value set no longer appear on the timeline. Was that intentional as I know there were some restrictions?
Previously and with M2O fields they would appear in category "UNASSIGNED".

File: web_timeline/static/src/js/timeline_renderer.js [hunk 4]
```diff
@@ -0,0 +1,633 @@
...
+         * Transform Odoo event object to timeline event object.
+         *
+         * @param {TransformEvent} evt
+         * @private
+         * @returns {Object}
+         */
+        event_data_transform: function (evt) {
+            const [date_start, date_stop] = this._get_event_dates(evt);
+            let group = evt[this.last_group_bys[0]];
+            if (group && group instanceof Array) {
```
Comment (chrisandrewmann) [inline web_timeline/static/src/js/timeline_renderer.js]: To fix issue where events are not displayed under UNASSIGNED group when using a M2M field.
Add this small change:

`            if (group && group instanceof Array && group.length > 0) {`

Comment (LudLaf) [reply web_timeline/static/src/js/timeline_renderer.js]: Suggested change `if (group && group instanceof Array && group.length > 0) {` added.

**Comment by CarlosRoca13** [comment]
Added fix and updated libraries as done in PR https://github.com/OCA/web/pull/2268

**Comment by pedrobaeza** [comment]
Have you tested? I find weird behaviors when testing it in v14.

**Comment by CarlosRoca13** [comment]
Yes, and I didn't found that problems...

**Pull request merged**